### PR TITLE
[FE-181] Refactor request detail modal

### PR DESCRIPTION
**Pull request by corovcam**
Solving issues/comments in #32 as well.

### CODE REVIEW

**Comment by corovcam** [comment]
> Thanks @corovcam, I have a suggestion that might simplify things quite a bit further (as we currently still battle with 3 different requests-related widgets displayed on record detail page) & make it occupy less space.
> 
> The proposal is to merge everything into a single box like this:
> 
> Screenshot 2024-07-02 at 15 01 46
> (just a raw mockup - feel free to adjust spacing / styling as needed)
> 
> With czech translations of "Žádosti" and "Čeká na vyřízení".
> 
> * First part contains a list of create request buttons available to current user
> * Second part contains a list of `is_open === true` requests
> * If there is no open request on a record, second part is completely omitted (including its header label) -- we assume users are not particularly interested in in requests they can do nothing about (they can look them up in dashboard if they want, though)
> 
> Does it sound reasonable to you? :)

Okay, thanks, yes that will simplify the list view by not including Accepted, Declined and Cancelled requests. One question, what if one have submitted some request waiting for approval (ViewOnlyModal without actions or Cancel modal with cancel action) and that one also has to approve/decline some other requests. The list would then not distinguish between Submitted by me and Submitted by someone else waiting for my approval (Requests to Approve), only when modal is opened one can see the difference. Of course, I can distinguish these two some other way (color, "To approve" label instead of "Submitted",...).

**Comment by mirekys** [comment]
> > Thanks @corovcam, I have a suggestion that might simplify things quite a bit further (as we currently still battle with 3 different requests-related widgets displayed on record detail page) & make it occupy less space.
> > The proposal is to merge everything into a single box like this:
> > Screenshot 2024-07-02 at 15 01 46
> > (just a raw mockup - feel free to adjust spacing / styling as needed)
> > With czech translations of "Žádosti" and "Čeká na vyřízení".
> > 
> > * First part contains a list of create request buttons available to current user
> > * Second part contains a list of `is_open === true` requests
> > * If there is no open request on a record, second part is completely omitted (including its header label) -- we assume users are not particularly interested in in requests they can do nothing about (they can look them up in dashboard if they want, though)
> > 
> > Does it sound reasonable to you? :)
> 
> Okay, thanks, yes that will simplify the list view by not including Accepted, Declined and Cancelled requests. One question, what if one have submitted some request waiting for approval (ViewOnlyModal without actions or Cancel modal with cancel action) and that one also has to approve/decline some other requests. The list would then not distinguish between Submitted by me and Submitted by someone else waiting for my approval (Requests to Approve), only when modal is opened one can see the difference. Of course, I can distinguish these two some other way (color, "To approve" label instead of "Submitted",...).



> > Thanks @corovcam, I have a suggestion that might simplify things quite a bit further (as we currently still battle with 3 different requests-related widgets displayed on record detail page) & make it occupy less space.
> > The proposal is to merge everything into a single box like this:
> > Screenshot 2024-07-02 at 15 01 46
> > (just a raw mockup - feel free to adjust spacing / styling as needed)
> > With czech translations of "Žádosti" and "Čeká na vyřízení".
> > 
> > * First part contains a list of create request buttons available to current user
> > * Second part contains a list of `is_open === true` requests
> > * If there is no open request on a record, second part is completely omitted (including its header label) -- we assume users are not particularly interested in in requests they can do nothing about (they can look them up in dashboard if they want, though)
> > 
> > Does it sound reasonable to you? :)
> 
> Okay, thanks, yes that will simplify the list view by not including Accepted, Declined and Cancelled requests. One question, what if one have submitted some request waiting for approval (ViewOnlyModal without actions or Cancel modal with cancel action) and that one also has to approve/decline some other requests. The list would then not distinguish between Submitted by me and Submitted by someone else waiting for my approval (Requests to Approve), only when modal is opened one can see the difference. Of course, I can distinguish these two some other way (color, "To approve" label instead of "Submitted",...).

In this situation, we kinda depend on curators that they know, they aren't on their own record (e.g. they got there using some link from a notification / from dashboard section on requests to approve), and if they are, they know which request they made (there wouldn't be >1-2 pending requests in most of cases), regular users would see just their own pending requests all the time. If this assumption becomes wrong, we could resort to distinguish mine vs. other requests using invenio rdm's approach (adding `is_current_user` to request metadata and visually tagging such requests as my own):
Screenshot 2024-07-02 at 16 01 48
Screenshot 2024-07-02 at 16 01 48

**Comment by corovcam** [comment]
> > > Thanks @corovcam, I have a suggestion that might simplify things quite a bit further (as we currently still battle with 3 different requests-related widgets displayed on record detail page) & make it occupy less space.
> > > The proposal is to merge everything into a single box like this:
> > > Screenshot 2024-07-02 at 15 01 46
> > > (just a raw mockup - feel free to adjust spacing / styling as needed)
> > > With czech translations of "Žádosti" and "Čeká na vyřízení".
> > > 
> > > * First part contains a list of create request buttons available to current user
> > > * Second part contains a list of `is_open === true` requests
> > > * If there is no open request on a record, second part is completely omitted (including its header label) -- we assume users are not particularly interested in in requests they can do nothing about (they can look them up in dashboard if they want, though)
> > > 
> > > Does it sound reasonable to you? :)
> > 
> > 
> > Okay, thanks, yes that will simplify the list view by not including Accepted, Declined and Cancelled requests. One question, what if one have submitted some request waiting for approval (ViewOnlyModal without actions or Cancel modal with cancel action) and that one also has to approve/decline some other requests. The list would then not distinguish between Submitted by me and Submitted by someone else waiting for my approval (Requests to Approve), only when modal is opened one can see the difference. Of course, I can distinguish these two some other way (color, "To approve" label instead of "Submitted",...).
> 
> > > Thanks @corovcam, I have a suggestion that might simplify things quite a bit further (as we currently still battle with 3 different requests-related widgets displayed on record detail page) & make it occupy less space.
> > > The proposal is to merge everything into a single box like this:
> > > Screenshot 2024-07-02 at 15 01 46
> > > (just a raw mockup - feel free to adjust spacing / styling as needed)
> > > With czech translations of "Žádosti" and "Čeká na vyřízení".
> > > 
> > > * First part contains a list of create request buttons available to current user
> > > * Second part contains a list of `is_open === true` requests
> > > * If there is no open request on a record, second part is completely omitted (including its header label) -- we assume users are not particularly interested in in requests they can do nothing about (they can look them up in dashboard if they want, though)
> > > 
> > > Does it sound reasonable to you? :)
> > 
> > 
> > Okay, thanks, yes that will simplify the list view by not including Accepted, Declined and Cancelled requests. One question, what if one have submitted some request waiting for approval (ViewOnlyModal without actions or Cancel modal with cancel action) and that one also has to approve/decline some other requests. The list would then not distinguish between Submitted by me and Submitted by someone else waiting for my approval (Requests to Approve), only when modal is opened one can see the difference. Of course, I can distinguish these two some other way (color, "To approve" label instead of "Submitted",...).
> 
> In this situation, we kinda depend on curators that they know, they aren't on their own record (e.g. they got there using some link from a notification / from dashboard section on requests to approve), and if they are, they know which request they made (there wouldn't be >1-2 pending requests in most of cases), regular users would see just their own pending requests all the time. If this assumption becomes wrong, we could resort to distinguish mine vs. other requests using invenio rdm's approach (adding `is_current_user` to request metadata and visually tagging such requests as my own): Screenshot 2024-07-02 at 16 01 48 Screenshot 2024-07-02 at 16 01 48

I see, okay, I'll push these small changes to this PR then.

File: oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/components/RequestListContainer.jsx
Comment (mirekys) [inline oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/components/RequestListContainer.jsx:60]: RequestList should pull itself everything it needs from context, avoid prop drilling

File: oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/components/RequestListContainer.jsx
Comment (mirekys) [inline oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/components/RequestListContainer.jsx:65]: RequestList should pull itself everything it needs from context, avoid prop drilling

**Review by mirekys — CHANGES_REQUESTED** [review]

File: oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/utils/hooks.js
Comment (mirekys) [inline oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/utils/hooks.js:37]: Please have a look here, if ConfirmationModal from oarepo-ui couldn't be used/adapted to needs: https://github.com/oarepo/oarepo-ui/blob/main/oarepo_ui/theme/assets/semantic-ui/js/oarepo_ui/forms/components/DeleteButton/DeleteButton.jsx#L18, to avoid creating multiple parallel implementations of the same concept

Comment (mirekys) [reply oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/utils/hooks.js:37]: Feel free to create PR on that component in oarepo-ui

File: oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/utils/hooks.js
Comment (mirekys) [inline oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/utils/hooks.js:42]: everything except `open` isn't really a state, as it is static once rendered, or isn't it? open/close state is already solved in hook from oarepo-ui

File: oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/utils/hooks.js
Comment (mirekys) [inline oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/utils/hooks.js:106]: add `request` to initializer props here

File: oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/utils/hooks.js
Comment (mirekys) [inline oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/utils/hooks.js:116]: this should be just private (non-exported) helper function

File: oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/utils/hooks.js
Comment (mirekys) [inline oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/utils/hooks.js:118]: Trying again later would never fix this issue of not providing an url :)

File: oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/utils/hooks.js
Comment (mirekys) [inline oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/utils/hooks.js:158]: replace this with more user-friendly methods, handling specific actions, either `acceptRequest(event)` or `doAction(REQUEST_TYPE)`. Methods should find out necessary callApi params from request type and request.links by itself

Comment (mirekys) [reply oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/utils/hooks.js:158]: Instead of doing this in my components using this client:
```jsx
    const { sendRequest } = useRequestsApi();  
  const acceptActionConfirmationHandler = () => confirmAction(() => onSubmit(() => sendRequest(request.links.actions?.accept, REQUEST_TYPE.ACCEPT)), REQUEST_TYPE.ACCEPT);
  //...
  return (request.links.actions?.accept && <Button title={i18next.t("Accept request")} onClick={acceptActionConfirmationHandler} positive icon labelPosition="left" floated="right" />
```

I'd like to be able to do:

```jsx
const { acceptRequest } = useRequestsApi(request);
return <Button title={i18next.t("Accept request")} onClick={acceptRequest} positive icon labelPosition="left" floated="right" />
```

Or slightly more generic:

```jsx
const { doAction } = useRequestsApi(request);
return <Button title={i18next.t("Accept request")} onClick={() => doAction(REQUEST_TYPE.ACCEPT)} positive icon labelPosition="left" floated="right" />
```

File: oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/components/RequestModal.jsx
Comment (mirekys) [reply oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/components/RequestModal.jsx:16]: could have another prop called `onSubmit`, that way a parent could provide its own callback to be called once modal content is submitted

File: oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/components/modals/CreateModal.jsx
Comment (corovcam) [reply oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/components/modals/CreateModal.jsx:69]: Each submit button also has `form` attribute to reference DOM form elements.

File: oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/utils/hooks.js
Comment (mirekys) [reply oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/utils/hooks.js:18]: Additionally, check https://github.com/oarepo/oarepo-requests/pull/34/files#r1662786416, I'm not sure if this useOnSubmit hook is even necessary...

**Comment by corovcam** [comment]
The last commit 3265bdd should fix the previous comments

**Comment by mirekys** [comment]
Hi @corovcam unable to build this:
ERROR in ./js/oarepo_requests_ui/record-requests/components/index.js 10:0-25
Module not found: Error: Can't resolve './modals' in '/nr-docs/.venv/var/instance/assets/js/oarepo_requests_ui/record-requests/components'
Did you miss the leading dot in 'resolve.extensions'? Did you mean '[".*",".js",".jsx"]' instead of '["*",".js",".jsx"]'?

**Review by mirekys — CHANGES_REQUESTED** [review]
Missing cs translation for "Pending" -> "Čeká na vyřízení"

File: oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/components/CreateRequestModalContent.jsx
Comment (mirekys) [inline oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/components/CreateRequestModalContent.jsx:31]: If a form has some custom fields, how do we submit field values here?

File: oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/components/RequestModal.jsx
Comment (corovcam) [reply oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/components/RequestModal.jsx:110]: Because not all props are used in the variable `ActionComponent`s. Some Actions need more props, some less. And since some components need to distinguish between `request` and `requestType` respectively, I also include additional prop. In other cases its just small redundancy. Also `request` and `requestType` have very similar interface in terms of `links`, so in many cases it doesn't matter what we use.

File: oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/utils/hooks.js
Comment (corovcam) [reply oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/utils/hooks.js:99]: Ahh yes, bad naming, but it works as expected.

File: oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/utils/hooks.js
Comment (corovcam) [reply oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/utils/hooks.js:31]: The thing is, we use one `Confirm` component from Semantic-UI in the entire `RequestModal` for all `ActionComponents`, which may or may not open the same (already rendered) `Confirm` modal. The props are set dynamically upon pressing the Action button to change the visual and logical layout of the Confirm modal.

File: oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/components/RequestModal.jsx
Comment (corovcam) [reply oarepo_requests/ui/theme/assets/semantic-ui/js/oarepo_requests_ui/record-requests/components/RequestModal.jsx:29]: Yes, some previous implementation had request types as the only context value passed down. Whole record is not really needed.

**Comment by corovcam** [comment]
The previous should be resolved now.

**Review by mirekys — APPROVED** [review]
LGTM now, good job! 👍

**Pull request merged**
